### ISSUE TITLE
Torch/onnx : Variable in dynamic axes can't be used in an if statement during the forward method in the model

### ISSUE THREAD

**Issue by Bahaabrougui** [pytorch/pytorch#58368]
**Describe the bug**
When specifying a variable appearing in the shape of the input as dynamic and not fixed, and if you're using that same variable in the forward function of the model to do some if statements and boolean tests, while exporting you get this warning : 
`TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!`.
Then when loading the model and trying to use it, you get an error of non matching shape as that variable is exported as a constant.

Example : Inputs are features extracted from audio files; input shape is (20, 9000, 40) where 20  is the batch size, 9000 is the sequence length (number of frames) and 40 is the number of features, or the model have been trained on sequence lengths of 198 but can generalize to larger sequence length and this is implemented in the forward() function.
Model was exported with opset_version = 13.

**Urgency**
There is actually a deadline for this university project I'm doing so it'd be great if this can be fixed or a solution can be provided, thanks in advance !

**System information**
- Torch version 1.8.1
- Linux Ubuntu 18.04
- ONNX Runtime installed from pip
- ONNX Runtime version: 1.7.0
- Python version: 3.6.9
- Visual Studio version (if applicable): Using PyCharm
- GCC/Compiler version (if compiling from source): -
- CUDA/cuDNN version: Using CPU
- GPU model and memory: -

**To Reproduce**
- Initialize a random input with a sequence length > 198 and pass it to the model, this is the code : 
`def to_numpy(tensor):`
    `return tensor.detach().cpu().numpy() if tensor.requires_grad else tensor.cpu().numpy()`

`x = torch.randn(20, 8000, 40, requires_grad=True)`
`ort_session = onnxruntime.InferenceSession(path_to_model)`
`ort_inputs = {ort_session.get_inputs()[0].name: to_numpy(inputs)}`
`ort_outs = ort_session.run(None, ort_inputs)` 
`predictions = ort_outs[0]`

- [model.zip](https://github.com/pytorch/pytorch/files/6492650/model.zip)

**Expected behavior**
Model should output a numpy array of predictions without displaying the error related to shape being different.

**Screenshots**
![image](https://user-images.githubusercontent.com/38329838/118397570-f5aa0300-b654-11eb-9b33-90e1075efd68.png)
This when running the model : 
![image](https://user-images.githubusercontent.com/38329838/118397647-5d604e00-b655-11eb-9a89-ae1043f5b770.png)

Thank you for your time and patience.

cc @houseroad @spandantiwari @lara-hdr @BowenBao @neginraoof @SplitInfinity

**Comment by BowenBao** [comment]
hi @Bahaabrougui, if statements needs to be scripted in order to be captured in the exported model. Checkout https://pytorch.org/docs/stable/onnx.html#tracing-vs-scripting for details.

**Comment by Bahaabrougui** [comment]
@BrowenBao thank you for you reply, I'll try it out.

**Comment by BowenBao** [comment]
closing. feel free to reopen if the issue persist even with scripting.
